src/library/searchquery.cpp

```diff
@@ -471,6 +471,221 @@ double DurationFilterNode::parse(const QString& arg, bool* ok) {
     return 60 * m + s;
 }
 
+// static
+constexpr double BpmFilterNode::kRelativeRangeDefault;
+
+// static
+double BpmFilterNode::s_relativeRange = kRelativeRangeDefault;
+
+// static
+void BpmFilterNode::setBpmRelativeRange(double range) {
+    // range < 0 would yield zero results because m_dRangeLow > m_dRangeHigh
+    VERIFY_OR_DEBUG_ASSERT(range >= 0) {
+        return;
+    }
+    s_relativeRange = range;
+}
+
+BpmFilterNode::BpmFilterNode(QString& argument, bool fuzzy, bool negate)
+        : m_fuzzy(fuzzy),
+          m_negate(negate),
+          m_isNullQuery(false),
+          m_isOperatorQuery(false),
+          m_isRangeQuery(false),
+          m_isHalfDoubleQuery(false),
+          m_operator("="),
+          m_bpm(0.0),
+          m_rangeLower(0.0),
+          m_rangeUpper(0.0),
+          m_bpmHalfLower(0.0),
+          m_bpmHalfUpper(0.0),
+          m_bpmDoubleLower(0.0),
+          m_bpmDoubleUpper(0.0) {
+    if (argument == kMissingFieldSearchTerm) {
+        m_isNullQuery = true;
+        return;
+    }
+    qWarning() << "     .BpmFilter:" << argument;
+
+    QRegularExpressionMatch opMatch = kNumericOperatorRegex.match(argument);
+    if (opMatch.hasMatch()) {
+        if (m_fuzzy) {
+            // fuzzy can't be combined with operators
+            return;
+        }
+        m_operator = opMatch.captured(1);
+        argument = opMatch.captured(2);
+        qWarning() << "     .opMatch:" << m_operator;
+    }
+
+    // Replace the locale's decimal separator with .
+    // This is handy if numbers are typed with the numpad.
+    argument.replace(',', '.');
+    bool isDouble = false;
+    const double bpm = argument.toDouble(&isDouble);
+    qWarning() << "     .isDouble:" << isDouble;
+    if (isDouble) {
+        if (m_fuzzy) {
+            qWarning() << "     .fuzzy";
+            // fuzzy search +- n%
+            m_rangeLower = floor((1 - s_relativeRange) * bpm);
+            m_rangeUpper = ceil((1 + s_relativeRange) * bpm);
+            m_isRangeQuery = true;
+        } else if (!opMatch.hasMatch() && !m_negate) {
+            qWarning() << "     .halfDouble";
+            // Simple 'bpm:NNN' search.
+            // Also searches for half/double matches.
+            // If decimals are provided, extend the core range, else search for
+            // exact matches.
+            ifDecimalsSetRange(argument, bpm);
+            // Include half/double BPM (rounded to int)
+            m_bpmHalfLower = floor(bpm / 2);
```

Comment:
it's a little awkward how some ranges are updated in the function as a side effect and others are updated manually.   And then, there is a lot of logic that switches depending on the mode of the query, which I believe is actually constant once this object is created, right? This creates a bit of difficulty following the logic with a lot of conditional branching.  

Based on the call pattern in this node, I think it might make more sense to use a different programming pattern: write a separate function for each search type that takes the constant parameters as currying arguments

So the constructor is basically:

```
switch (mode) {
 case operator_mode:
   m_matchFunc = match_operator_fn(op, bpm)
  case fuzzy_mode:
    m_matchFunc = match_fuzzy_fn(bpm)
  ...
}

bool BpmFilterNode::match(const TrackPointer& pTrack) const {
  // get the track bpm...
  return m_matchFunc(track_bpm);
}
```

and then 

```
my_matcher_fn_type match_operator_fn(string op, bpm compare_bpm) {
	// precalculate range and whatever

	// curry necessary arguments and return a customized function for this search
	return [=](float track_bpm) {
      // do stuff with range as compared to track_bpm
	};
}
```

This way the individual match functions can be easily tested in isolation and the ranges are still precalculated but don't require separate member variables to do so.